Light refactor of the heavily refactored statement classes

paul-rogers opened the pull request:
PR #12845 converted `SqlLifecycle` into a set of statement classes. The `DirectStatement` class combined planning and execution into a single `execute()` method. However, it turns out that a consumer outside of the Druid code base had need to separate the plan portion of the operation from execution.

This PR provides that separation in the form of a `ResultSet` class that holds onto the "physical plan" (such as it is in Druid) and does the execution step. The previous `execute()` method still exists for the code, within Druid, that uses the all-in-one approach.

<hr>

This PR has:
- [X] been self-reviewed.
- [X] added Javadocs for most classes and all non-trivial methods. Linked related entities via Javadoc links.
- [X] added comments explaining the "why" and the intent of the code wherever would not be obvious for an unfamiliar reader.
- [ ] been tested in a test Druid cluster.

rohangarg commented on sql/src/main/java/org/apache/druid/sql/DirectStatement.java:
```diff
@@ -62,8 +65,81 @@ public class DirectStatement extends AbstractStatement implements Cancelable
+   * Represents the execution plan for a query with the ability to run
+   * that plan (once).
+   */
+  public class ResultSet implements Cancelable
```
is a name like `PlannerResultWrapper` or something similar more suitable? because this class seems to augment the functionalities of `PlannerResult` with some statement related accessors.

paul-rogers replied:
The name echos the JDBC concept of a result set. It is a way to get results for a statement. The idea, in JDBC, is that rather than the statement having two "modes" (running or not), the lifecycle is encoded in the class structure. A statement does the work of everything up to a set of results; the result set then manages those results. As in JDBC, some operations are common (close, etc.)

This is needed because of a use case (outside of Druid itself) where we want to manage the results in a different thread than the one that prepares the statement. Splitting out the functionality to a `ResultSet` is a bit cleaner than having "is prepared and ready to run" state in the statement itself.

This is a quick & dirty refactoring to solve issues created by the previous PR. As you point out, it could be improved after we give it a bit more thought.

rohangarg replied:
Thanks a lot for explanation! Yes, I agree that the semblence to JDBC concept makes sense and I was also thinking in the same line. But I was thinking that maybe then the `run` or `execute` method should be a part of the statement instead of the resultset. Maybe, the result-set could start with a `Yielder` object over the result sequence.

But agree that all that could be done after more thought as a follow up. And completely agree with the concept of breaking Statement into Statement and ResultRet :+1:

rohangarg reviewed (approved):
LGTM % comments + CI (intellij inspect is legit, the code coverage one might get fixed if more `ResultSet` methods are called in `SqlResourceTest`)

paul-rogers commented:
Fixed one actual issue. Odd that the UT worked on Java 8 but not on the other versions: the issue seemed unrelated to the Java version.

`SegmentMetadataCacheTest.testLocalSegmentCacheSetsDataSourceAsGlobalAndJoinable` appears flaky, it didn't fail in the IDE. Added a longer timeout as a possible workaround. The two IT failures appear infrastructure-related. Pushing a new commit to fix the actual issues and retry the flaky ITs.

rohangarg reviewed (approved):
minor nits, can be picked up in follow ups

paul-rogers commented:
Good point on the semantics. We're a bit locked into the fact that "plan" in Druid means "create a plan and return a sequence backed by a bunch of query runners that just about have started the query." What we want is to plan up to the native query, then have a separate step that runs that native query. We can sort that out later.
